ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DiffSeq.hs

```diff
@@ -0,0 +1,281 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+
+{-# LANGUAGE UndecidableInstances       #-}
+
+module Ouroboros.Consensus.Storage.LedgerDB.HD.DiffSeq (
+    -- * Sequences of diffs
+    DiffSeq (..)
+  , Element (..)
+  , InternalMeasure (..)
+  , Length (..)
+  , RootMeasure (..)
+  , SlotNo (..)
+    -- * Short-hands for type-class constraints
+  , SM
+    -- * API: derived functions
+  , cumulativeDiff
+  , extend
+  , extend'
+  , length
+  , mapDiffSeq
+  , maxSlot
+  , splitl
+  , splitlAt
+  , splitlAtFromEnd
+  , splitr
+  , splitrAt
+  , splitrAtFromEnd
+  ) where
+
+import           Prelude hiding (length, splitAt)
+
+import qualified Control.Exception as Exn
+import           Data.Bifunctor (Bifunctor (bimap))
+import           Data.Group
+import           Data.Monoid (Sum (..))
+import           Data.Semigroup (Max (..))
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+
+import           Data.FingerTree.RootMeasured.Strict hiding (split, splitl,
+                     splitlr, splitr)
+import qualified Data.FingerTree.RootMeasured.Strict as RMFT (splitlr)
+import           Data.Map.Diff.Strict (Diff)
+
+import qualified Cardano.Slotting.Slot as Slot
+
+{-------------------------------------------------------------------------------
+  Sequences of diffs
+-------------------------------------------------------------------------------}
+
+newtype DiffSeq k v =
+  DiffSeq
+    (StrictFingerTree
+      (RootMeasure k v)
+      (InternalMeasure k v)
+      (Element k v)
+    )
+  deriving stock (Generic, Show, Eq)
+  deriving newtype (Semigroup, Monoid)
+  deriving anyclass (NoThunks)
+
+
+-- The @'SlotNo'@ is not included in the root measure, since it is
+-- not a @'Group'@ instance.
+data RootMeasure k v = RootMeasure {
+    -- | Cumulative length
+    tmLength :: {-# UNPACK #-} !Length
+    -- | Cumulative diff
+  , tmDiff   :: !(Diff k v)
+  }
+  deriving stock (Generic, Show, Eq, Functor)
+  deriving anyclass (NoThunks)
+
+data InternalMeasure k v = InternalMeasure {
+    -- | Cumulative length
+    imLength :: {-# UNPACK #-} !Length
+    -- | Right-most slot number
+  , imSlotNo ::                !(Maybe SlotNo)
+  }
+  deriving stock (Generic, Show, Eq, Functor)
+  deriving anyclass (NoThunks)
+
+data Element k v = Element {
+    elSlotNo :: {-# UNPACK #-} !SlotNo
+  , elDiff   ::                !(Diff k v)
+  }
+  deriving stock (Generic, Show, Eq, Functor)
+  deriving anyclass (NoThunks)
+
+
+-- | Length as a @'Sum'@.
```

Comment:
Would it make sense to explain what this length refers to? Eg

```haskell
Length of the sequence of differences
```

I don't get the "as a Sum" part. Does it mean that its Group instance is defined through the Group instance for `Sum`?

Reply:
It means that `Length` is, with respect to group properties, equivalent to a `Sum`. This does not have to be mentioned in this `haddock` comment however, so I have removed it.

```haskell
-- | Length of a sequence of differences.
newtype Length = Length {unLength :: Int}
```